File: distributed-tracing/ruby-rails-app/delivery/app/api/delivery.rb
require 'securerandom'

# Driver for Delivery service which manages the delivery of shirts.
#
# @author Yogesh Prasad Kurmi (ykurmi@vmware.com).
#
class Delivery < Grape::API
    format :json
    resource :dispatch do
        desc "Dispatch the order"
        use Rack::Tracer
        post '/:order_num' do
          sleep(1)
          order_num = params[:order_num]
          if rand(1..5) == 5
            message = "Random Service Unavailable!"
            Rails.logger.warn(message)
            env['rack.span'].set_tag('error', true)
            status 503
            return message
          end
          if rand(1..10) == 10
            order_num = nil
          end

          if !order_num
            message = "Invalid Order Num!"
            Rails.logger.warn(message)
            env['rack.span'].set_tag('error', true)
            status 400
            return message
          end
          packed_shirts = nil
          if rand(1..10) !=  10
            packed_shirts = params[:shirts]
          end

          if !packed_shirts
            message = "No shirts to deliver!"
            Rails.logger.warn(message)
            env['rack.span'].set_tag('error', true)
            status 400
            return message
          end

          tracking_num = SecureRandom.uuid
          status 200
          {"orderNum": order_num,
           "trackingNum": tracking_num,
           "status": "shirts delivery dispatched"}
        end
  
    end
    resource :retrieve do
      desc "Retrieve Order"
      use Rack::Tracer
      post '/:order_num' do
        sleep(1)
        order_num = params[:order_num]
        if !order_num
          message = "Invalid Order Num!"
          Rails.logger.warn(message)
          env['rack.span'].set_tag('error', true)
          status 400
          return message
        end
        status 200
        {"status": "Order:" + order_num + " returned"}
      end
    end
  end

File: distributed-tracing/go-app/services/delivery/delivery.go
// 2019. Mahesh Voleti (mvoleti@vmware.com)

package delivery

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/go-chi/chi"
	opentracing "github.com/opentracing/opentracing-go"
	otrext "github.com/opentracing/opentracing-go/ext"

	. "wavefront.com/hackathon/beachshirts"
)

type DeliveryServer struct {
	HostURL       string
	Router        *chi.Mux
	DeliveryQueue chan PackedShirts

	tracer opentracing.Tracer
}

func NewServer() *DeliveryServer {
	r := chi.NewRouter()
	deliveryQueue := make(chan PackedShirts, GlobalConfig.DeliveryQueueSize)

	server := &DeliveryServer{GlobalConfig.DeliveryHost, r, deliveryQueue, opentracing.GlobalTracer()}

	r.Route("/delivery", func(r chi.Router) {
		r.Post("/dispatch/{orderNum}", server.dispatch)
		r.Get("/return/{orderNum}", server.retrieve)
	})

	return server
}

func (s *DeliveryServer) Start() error {
	log.Println("Delivery Server listening @", s.HostURL)

	ticker := time.NewTicker(30 * time.Second)
	go func() {
		for _ = range ticker.C {
			span := s.tracer.StartSpan("processDispatch")

			log.Printf("Processing %d entries in the dispatch queue!\n", len(s.DeliveryQueue))
			done := false
			for !done {
				select {
				case item := <-s.DeliveryQueue:
					log.Printf("\t%d shirts delivered\n", len(item.Shirts))
				default:
					done = true
				}
			}

			span.Finish()
		}
	}()

	return http.ListenAndServe(s.HostURL, s.Router)
}

func (s *DeliveryServer) dispatch(w http.ResponseWriter, r *http.Request) {
	span := NewServerSpan(r, "dispatch")
	defer span.Finish()

	RandSimDelay()

	if RAND.Float32() < GlobalConfig.SimFailDelivery1 {
		otrext.Error.Set(span, true)
		WriteError(w, "Failed to dispatch shirts!", http.StatusServiceUnavailable)
		return
	}

	orderNum := chi.URLParam(r, "orderNum")

	if RAND.Float32() < GlobalConfig.SimFailDelivery2 {
		orderNum = ""
	}

	if orderNum == "" {
		otrext.Error.Set(span, true)
		WriteError(w, "Invalid Order Num", http.StatusBadRequest)
		return
	}

	var packedShirts PackedShirts
	json.NewDecoder(r.Body).Decode(&packedShirts)

	if RAND.Float32() < GlobalConfig.SimFailDelivery3 {
		packedShirts = PackedShirts{}
	}

	if packedShirts.Shirts == nil || len(packedShirts.Shirts) == 0 {
		otrext.Error.Set(span, true)
		WriteError(w, "No shirts to deliver", http.StatusBadRequest)
		return
	}

	s.DeliveryQueue <- packedShirts

	trackingNum := NewOrderNum()
	log.Printf("Tracking number of Order: %s is %s\n", orderNum, trackingNum)

	RandSimDelay()

	status, _ := json.Marshal(DeliveryStatus{OrderNum: orderNum, TrackingNum: trackingNum, Status: "shirts delivery dispatched"})
	w.Write(status)
}

func (s *DeliveryServer) retrieve(w http.ResponseWriter, r *http.Request) {
	span := NewServerSpan(r, "retrieve")
	defer span.Finish()

	orderNum := chi.URLParam(r, "orderNum")
	if orderNum == "" {
		otrext.Error.Set(span, true)
		WriteError(w, "Invalid Order Num", http.StatusBadRequest)
	}
	fmt.Fprintf(w, "Order: %s returned\n", orderNum)
}
